firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/datagrams.txt

```diff
@@ -0,0 +1,474 @@
+    
+    WHEN("A body with a complete 2-byte header, and a non-empty paylaod buffer consistent with the length field of the header is parsed")
+      THEN("The parse method reports ok status")
+      THEN("After the parse method is called, The value returned by the seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The payload returned from the payload accessor method is equal to the payload from the input_buffer body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the Datagram constructor") 
+      THEN("the seq accessor method returns a reference to the sequence given in the Datagram constructor")
+      THEN("the payload given in the Datagram constructor is independent of the input buffer body")
+
+    WHEN("A body with a complete 2-byte header, and a non empty payload buffer consistent with the length field of the header, is altered after it's parsed")
+      THEN("The parse method reports ok status")
+      THEN("After the parse method is called, The value returned by the seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the length accessor method is equal to the length field of the original input_buffer body's header")
+      THEN("The payload buffer returned from the payload accessor method is equal to the original payload from the body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the Datagram constructor") 
+      THEN("the seq accessor method returns a reference to the sequence given in the Datagram constructor")
+      THEN("the payload given in the Datagram constructor is independent of the input buffer body")
+
+    WHEN("A body with a complete 2-byte header, and paylaod buffer as '0x00' consistent with the length field of the header is parsed")
+      THEN("The parse method reports ok status")
+      THEN("After the parse method is called, The value returned by the seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The payload returned from the payload accessor method is equal to the payload from the input_buffer body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the Datagram constructor") 
+      THEN("the seq accessor method returns a reference to the sequence given in the Datagram constructor")
+      THEN("the payload given in the Datagram constructor is independent of the input buffer body")
+
+    WHEN("A body with a complete 2-byte header, and a non-empty payload buffer where the length field of the header is inconsistent with the actual length of the payload buffer")
```

Comment:
This WHEN and the next WHEN need to specify that the parse method gets called. It might be awkward to put it at the end of the sentence for this WHEN, so you could change the order: WHEN the parse method is called on a body with ...